### PR TITLE
DOC: mention string, bytes, and void dtypes in dtype intro

### PR DESCRIPTION
**Pull request by ngoldbaum**
This is part of my work on docs for stringdtype. I'm issuing this separately from the stringdtype PR to ease review.

I moved the section comparing python and C data type names to later in the document, after material that made more sense to me at the beginning of the document.

I also added some discussion about working with string or bytes data. I mentioned some common footguns working with the bytes dtype and also explicitly mentioned the unstructured void dtype as an alternative option. As far as I can tell, unstructured void dtypes are not mentioned anywhere in the docs yet but they seem like a natural thing to mention here.

EDIT: also made sure the dtype classes all show up in the API docs so they can be cross-referenced

### CODE REVIEW

File: doc/source/reference/routines.other.rst [hunk 1]
```diff
@@ -33,4 +33,10 @@ Utility
    show_runtime
    broadcast_shapes
 
-.. automodule:: numpy.dtypes
+DType classes and utility (:mod:`numpy.dtypes`)
```
Comment (mhvk) [inline doc/source/reference/routines.other.rst:36]: utility -> utilities?

File: doc/source/user/basics.types.rst [hunk 2]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
 NumPy supports a much greater variety of numerical types than Python does.
 This section shows which are available, and how to modify an array's data-type.
 
+NumPy numerical types are instances of `numpy.dtype` (data-type) objects,
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:15]: `instance` has specific meaning in python: I would expect `type(np.int8)` now to be `np.dtype`, but this is not true - they all derive from `np.generic` I don't think it is helpful to suggest it is.

EDIT: darn, I see that this is simply all from the previous piece of text...

File: doc/source/user/basics.types.rst [hunk 3]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
...
+
+Data-types can be used as functions to convert python numbers to array scalars
+(see the array scalar section for an explanation), python sequences of numbers
+to arrays of that type, or as arguments to the dtype keyword that many numpy
+functions or methods accept. Some examples::
+
+    >>> x = np.float32(1.0)
+    >>> x
+    1.0
+    >>> y = np.int_([1,2,4])
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:28]: I'm not sure we should suggest such non-standard way to initialize an array here (at least, I've not seen this done in any code I've come across...)

File: doc/source/user/basics.types.rst [hunk 4]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
 NumPy supports a much greater variety of numerical types than Python does.
 This section shows which are available, and how to modify an array's data-type.
 
+NumPy numerical types are instances of `numpy.dtype` (data-type) objects,
+each having unique characteristics.  Once you have imported NumPy using ``import
+numpy as np`` the dtypes are available as e.g. `numpy.bool`,
+`numpy.float32`, etc.
+
+Data-types can be used as functions to convert python numbers to array scalars
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:20]: Reading this sentence again feels super misleading. This is not a data type! E.g., one cannot indicate byte order with it.

Also, do we even want to encourage creation of scalars? I think we should go straight to talking about arrays.

File: doc/source/user/basics.types.rst [hunk 5]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
...
+
+Array types can also be referred to by character codes, mostly to retain
+backward compatibility with older packages such as Numeric, but also for
+defining :ref:`structured data types <structured_arrays>`.  Some
+documentation may refer to these character codes, for example::
+
+  >>> np.array([1, 2, 3], dtype='f')
+  array([1.,  2.,  3.], dtype=float32)
+
+We recommend using dtype objects instead.
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:43]: Briefly tell why (and, for me, how to do strings and different byte order in that case...)

File: doc/source/user/basics.types.rst [hunk 6]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
...
+
+To convert the type of an array, use the .astype() method (preferred) or
+the type itself as a function. For example: ::
+
+    >>> z.astype(float)                 #doctest: +NORMALIZE_WHITESPACE
+    array([0.,  1.,  2.])
+    >>> np.int8(z)
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:50]: For a `basics` document, no reason to give another way.

File: doc/source/user/basics.types.rst [hunk 7]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
...
+    array([0, 1, 2], dtype=int8)
+
+Note that, above, we use the *Python* float object as a dtype.  NumPy knows that
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:53]: Again, I'd suggested keeping it to the recommended way first and then perhaps have a separate section on alternatives/shortcuts.

File: doc/source/user/basics.types.rst [hunk 8]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
...
+    >>> z.dtype
+    dtype('uint8')
+
+dtype objects also contain information about the type, such as its bit-width
+and its byte-order.  The data type can also be used indirectly to query
+properties of the type, such as whether it is an integer::
+
+    >>> d = np.dtype(int)
+    >>> d #doctest: +SKIP
+    dtype('int32')
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:70]: Might as well change this to `int64` since now even on windows that will always be true.

Or just change to float!

File: doc/source/user/basics.types.rst [hunk 9]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
...
+  >>> a
+  array([b'\x68\x65\x6C\x6C\x6F\x00\x00', b'\x77\x6F\x72\x6C\x64\x00\x00'],
+        dtype='|V7')
+  >>> a[0] == np.void(x[0])
+  True
+
+Advanced types, not listed above, are explored in section
+:ref:`structured_arrays`.
+
+Relationship Between NumPy Data Types and C Data Data Types
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:152]: You could have a `Reletaionship Between Numpy and Python Data Types here`.

**Review by mhvk — COMMENTED** [review]
Hmm, when I started reviewing this, I thought what a strange additions you made (see in-line comments), but then I realized that was all in the piece of text you moved... Your actual additions I think are good!

Of course, that moved piece also led me to the same problems as in https://github.com/numpy/numpy/issues/25508#issuecomment-1871469944 -- recommending against character codes when most users will quite quickly encounter them anyway (and *have* to understand them) seems rather backward to me.

So, bottom line is that I'm not sure what is best - ideally, we have an actual complete (and ideally concise) system of defining dtypes in place before we change the documentation to something that is incomplete and less helpful, as was done to this file already before your PR...

**Comment by ngoldbaum** [comment]
@mhvk would you mind if we leave the text that was already there as-is and just review the new text I added? And maybe file an issue to come back to the introductory text?

File: doc/source/user/basics.types.rst (hunk 9 above)
Comment (ngoldbaum) [reply doc/source/user/basics.types.rst:152]: That's true, but I'd prefer not to do that right now, if that's OK.

File: doc/source/user/basics.types.rst [hunk 10]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
...
+    >>> d
+    dtype('int64')
+
+    >>> np.issubdtype(d, np.integer)
+    True
+
+    >>> np.issubdtype(d, np.floating)
+    False
+
+Numerical Data Types
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:78]: This is the start of the new bit, right?

I like the first paragraph about numerical types a lot. And I do think it makes sense to have it after the basic examples.

File: doc/source/user/basics.types.rst [hunk 11]
```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
...
+`numpy.intp`, have differing bitsizes, dependent on the platforms
+(e.g. 32-bit vs. 64-bit CPU architectures).  This should be taken into account
+when interfacing with low-level code (such as C or Fortran) where the raw memory
+is addressed.
+
+Data Types for Strings and Bytes
+--------------------------------
+
+In addition to numerical types, NumPy also supports storing unicode strings, via
+`numpy.dtypes.StrDType` (``U`` character code), null-terminated byte sequences via
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:96]: I guess my main comment on your added text is that, since the `np.dtypes...DType` classes are as yet not directly useful, I would not mention them here at all.

**Review by mhvk — COMMENTED** [review]
OK, a review just of what you added: I like it, but would not mention the long names if you are not going to actually use them.

But I think when you move a piece so much up front, it should actually be good... I suggest improving it by shortening: just show a few basic array instantiation calls, ignoring the scalars part altogether - it is irrelevant here. You could even mostly just let the array constructor itself decide on float or int, to avoid passing in anything, and then perhaps give a single example with an overflow (for which I would use a character code, but that's me...)

**Comment by ngoldbaum** [comment]
> since the np.dtypes...DType classes are as yet not directly useful

They are directly useful right now, FWIW:

```
In [4]: np.array(["hello", "world"], dtype=np.dtypes.StrDType(10))
Out[4]: array(['hello', 'world'], dtype='<U10')
```

But point taken, let's not move to using those names in this one place. I also edited down the introductory text a bit to remove the most objectionable stuff.

File: doc/source/user/basics.types.rst (hunk 7 above)
Comment (ngoldbaum) [reply doc/source/user/basics.types.rst]: Kept this but reworded it a bit so the example uses `float64` initially but subsequently explains that we could have used the python builtin. This is common enough I think it's worth explaining here.

File: doc/source/user/basics.types.rst (hunk 4 above)
Comment (ngoldbaum) [reply doc/source/user/basics.types.rst]: I reworded this a bit and now refer to these objects as scalar types (which they are) but don't explain exactly what that means when I introduce the term, since I think that would be distracting. I also added a cross-reference to the dtype reference docs so there's at least a hyperlink to more in-depth docs.

File: doc/source/reference/routines.other.rst (hunk 1 above)
Comment (mhvk) [reply doc/source/reference/routines.other.rst:36]: Is there a `utility` yet? If not, maybe just `DType classes`?

File: doc/source/user/basics.types.rst [hunk 12]
```diff
@@ -12,6 +12,138 @@ Array types and conversions between types
...
+  >>> np.array([1, 2, 3], dtype='f')
+  array([1.,  2.,  3.], dtype=float32)
+  >>> np.array([1, 2, 3], dtype='d')
+  array([1.,  2.,  3.], dtype=float64)
+
+See :ref:`arrays.dtypes.constructing` for more information about specifying and
+constructing data type objects, including how to specify parameters like the
+byte order.
+
+To convert the type of an array, use the .astype() method (preferred) or
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:38]: I'd still stop the sentence after "method" - no need for introducing the "type itself as a function"

File: doc/source/user/basics.types.rst [hunk 13]
```diff
@@ -12,6 +12,138 @@ Array types and conversions between types
...
+the type itself as a function. For example: ::
+
+    >>> z.astype(np.float64)                 #doctest: +NORMALIZE_WHITESPACE
+    array([0.,  1.,  2.])
+
+Note that, above, we could have used the *Python* float object as a dtype
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:44]: This works nicely!

File: doc/source/user/basics.types.rst [hunk 14]
```diff
@@ -12,6 +12,138 @@ Array types and conversions between types
...
+
+Data Types for Strings and Bytes
+--------------------------------
+
+In addition to numerical types, NumPy also supports storing unicode strings, via
+the `numpy.str_` dtype (``U`` character code), null-terminated byte sequences via
+`numpy.bytes_` (``S`` character code), and arbitrary byte sequences, via
+`numpy.void` (``V`` character code).
+
+The ``StrDType``, ``BytesDType``, and ``VoidDType`` are *fixed-width* data
```
Comment (mhvk) [inline doc/source/user/basics.types.rst:92]: Left-over?  Replace with "All the above are ..."

File: doc/source/user/basics.types.rst (hunk 9 above)
Comment (mhvk) [reply doc/source/user/basics.types.rst:144]: I think it is less necessary now - and in any case fine not to do it in this PR.

File: numpy/dtypes.py [hunk 15]
```diff
@@ -49,8 +46,8 @@
     * - Complex
       - ``Complex64DType``, ``Complex128DType``, ``CLongDoubleDType``
 
-    * - Strings
-      - ``BytesDType``, ``BytesDType``
+    * - Strings and Bytestrings
+      - ``StrDType``, ``BytesDType``
```
Comment (mhvk) [inline numpy/dtypes.py:50]: Nice catch!

**Review by mhvk — COMMENTED** [review]
Thanks so much, this now reads very well. Only a few nits left.

**Comment by ngoldbaum** [comment]
Responded to latest comments, should be good to go now. Maybe @mdhaber or @rossbar want to look this over?

One wrinkle I noticed looking at the docs build, this introduces some new reference warnings that I can't easily fix:

```
/Users/goldbaum/Documents/numpy/build-install/usr/local/lib/python3.11/site-packages/numpy/dtypes.py:docstring of numpy.dtypes.BytesDType:87: WARNING: autosummary: stub file not found 'numpy.dtypes.BytesDType.type'. Check your autosummary_generate setting.
```

And similar warnings for all of the other DType classes. I think that's happening is that both the dtype metaclass and dtype class share a python `type` attribute, and the way we've implemented the python inheritance for this in C (perhaps also somehow combined with the use of `T_OBJECT` to define the attribute) is confusing sphinx. If I change dtypemeta to use a `scalar_type` python-level name, that fixes it, but that seems like a more invasive risky change that I want to take on in this PR. Ping @seberg in case you weren't aware of this issue with the layering of the dtype API.

**Review by mhvk — APPROVED** [review]
Looks all good as far as I'm concerned!

**Review by matheussouza88 — APPROVED** [review]

**Comment by ngoldbaum** [comment]
Pulling this in, if anyone has any edits they didn't catch before merging I'm happy to followup.

File: doc/source/reference/routines.rst [hunk 16]
```diff
@@ -24,6 +24,7 @@ indentation.
    routines.ctypeslib
    routines.datetime
    routines.dtype
+   routines.dtypes
```
Comment (rgommers) [inline doc/source/reference/routines.rst:27]: @ngoldbaum this file doesn't actually exist. Did you forget to commit it, or is this some leftover?

Comment (ngoldbaum) [reply doc/source/reference/routines.rst:27]: Hmm, that's odd, must have forgotten to `git add` it back when I made this PR. Will push a fix tomorrow.
